common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/Indexer.scala

```diff
@@ -25,14 +24,77 @@ trait Indexable[T] {
+/**
+  * Versioning:
+  *
+  * When the merger makes the decision to merge some works, it modifies
+  * the content of the affected works. Despite the content of these works
+  * being modified, their version remains the same. However, the merger
+  * attaches the number of merged sources (`nMergedSources`) to the target
+  * work and may choose to redirect some of the source works.
+  *
+  * When we ingest those works, we need to make sure that the merger
+  * modified works are never overridden by unmerged works for the same
+  * version still running through the pipeline.
+  *
+  * We also need to make sure that, if a work is modified by a source in
+  * such a way that it shouldn't be merged (or redirected) anymore, the
+  * new unmerged version is ingested and never replaced by the previous
+  * merged version still running through the pipeline.
+  *
+  * We can do that by ingesting works into Elasticsearch with a document
+  * version derived from a combination of work version, number of merged
+  * sources and whether it is visible, redirected or invisible.
+  *
+  * More specifically, by multiplying the work version by 10, we make sure
+  * that a new version of a work always wins over previous versions
+  * (merged or unmerged, provided that there are fewer than 10 merged
+  * sources, which we think is a reasonable assumption).
+  *
+  * We make sure that a merger modified work always wins over other works
+  * with the same version, by adding the number of merged sources to
+  * work.version * 10.
+  *
+  * The exact same logic is applied to image versions, using both the
+  * merge state and the version numbers of the image's SourceWorks.
+  *
+  *  Gotchas:
+  *
+  *  We don't believe this causes us issues right now, but it's worth noting
+  *  that an unlinked work with the same version wouldn't override its merged
+  *  predecessor (even if it was then linked to another, different work).
+  *
+  *   Image version and image source version have the same precedence. If this
+  *   became an issue we'd probably want to do something like:
+  *
+  *   def version(image: AugmentedImage) =
+  *     10 * (2*image.version + image.source.version)
+  */
+object Indexable extends Logging {
 
   implicit val imageIndexable: Indexable[AugmentedImage] =
     new Indexable[AugmentedImage] {
       def id(image: AugmentedImage) =
         image.id.canonicalId
+
       def version(image: AugmentedImage) =
-        image.version
+        10 * (image.version + image.source.version) +
+          (warnOnTooManySources(image).source match {
+            case SourceWorks(_, Some(redirected), nMergedSources) =>
+              nMergedSources + 1
+            case SourceWorks(_, None, nMergedSources) => nMergedSources
+          })
+
+      private def warnOnTooManySources(
+        image: AugmentedImage): AugmentedImage = {
+        image.source match {
+          case SourceWorks(_, _, nMergedSources) if nMergedSources >= 10 =>
+            warn(
```

Comment:
Thinking about this again: should this just be a warning log? To go see warnings we need to know that there's something to take a look at, and that would be because points out to us that the data in the API is inconsistent. I'm not sure if that's what we want. I'm tempted to say make the message fail if `nMergedSources >= 10`, cause at this point we can't guarantee data consistentcy.

Reply:
Good point, will change that